### PR TITLE
2.0.0 dev singeleton mqtt

### PR DESCRIPTION
**Pull request by SmbatYeranyan**

### CODE REVIEW

File: entities/Device.js [hunk 1]
```diff
@@ -1,85 +1,132 @@
...
+            throw new Error('Provide the userId during instantiation');
+        } else if (!(patch)) {
+            throw new Error('Provide a patch of parameters to update');
+        } else if (!(Object.keys(patch).length)) {
+            throw new Error('Provide a patch with some parameters to update');
+        }
+
+        for (var x in patch) {
+            if (!(this.hasOwnProperty(x))) {
+                throw new Error('Provide a patch with relevant parameters to update');
```
Comment (jontore) [inline entities/Device.js]: So you can not patch attributes that are not properties on the device.

File: entities/Device.js [hunk 2]
```diff
@@ -1,85 +1,132 @@
...
+            }
+        }
+
+        return new Promise((resolve, reject) => {
+            this.ajax.patch(`/devices/${this.deviceId}`, patch, raw)
+                .then((response) => {
+                    this.name = response.name;
```
Comment (jontore) [inline entities/Device.js]: Maybe this can be extracted to a method, since this is the same as in the constructor.

File: entities/Device.js [hunk 3]
```diff
@@ -1,85 +1,132 @@
...
+                    //right now the object hangs around, but on the cloud it is gone
+                    resolve(response)
+                }).catch((error) => {
+                    reject(error);
+                });
+        })
+    }
+    sendCommand(command, raw) {
+        if (!(this.deviceId)) {
+            throw new Error('Provide the userId during instantiation');
```
Comment (jontore) [inline entities/Device.js]: userId?

File: entities/Device.js [hunk 4]
```diff
@@ -1,85 +1,132 @@
...
     }
 
-    return new Promise((resolve, reject) => {
-      this.ajax.patch(`/devices/${this.deviceId}`, patch, raw)
-        .then((response) => {
-          resolve(response);
-        }).catch((error) => {
-          reject(error);
+    connect(transport) {
+        if (!transport) {
```
Comment (jontore) [inline entities/Device.js]: Here you can use es6 default property ```connect(transport = 'mqtt') ```

File: entities/Device.js [hunk 5]
```diff
@@ -1,85 +1,132 @@
...
+            transport = "mqtt"
+        }
+
+        let connection = new Connection();
+        let getChannel = this.getChannel();
+        let subscribeMqtt = new Promise((resolve, reject) => {
+            //if (!eventCallback) throw Error("You must provide a callback for data")
```
Comment (jontore) [inline entities/Device.js]: Out commented line.

Comment (jontore) [inline entities/Device.js]: Can this be done at the same time, looks like you should first wait for the cahnnel to be resolved before opening a connection.

File: test/device.test.js [hunk 6]
```diff
@@ -6,223 +10,286 @@ import sinonChai from 'sinon-chai';
...
+            owner: 'fakeOwner',
+            openToPublic: false,
+            ajax: {
+                url: 'fakeURL',
+                token: '12345',
+                tokenType: 'Bearer'
+            }
+        };
+
+        // deviceStub = {
```
Comment (jontore) [inline test/device.test.js]: This can probably be deleted.

File: test/device.test.js [hunk 7]
```diff
@@ -6,223 +10,286 @@ import sinonChai from 'sinon-chai';
...
+
+        it('should throw an error if no deviceId given to look up', function() {
+            deviceInstance.deviceId = null;
+            var fn = function() {
+                deviceInstance.updateDevice();
+            };
+            expect(fn).to.throw(Error);
+        });
+
+        it('should give a body of parameters at all to update', function() {
```
Comment (jontore) [inline test/device.test.js]: These descriptions are not that descriptive. We're expecting it to throw if these conditions are not met. Maybe wrap it in a describe block ```describe('throws an error when,```

File: test/device.test.js [hunk 8]
```diff
@@ -6,223 +10,286 @@ import sinonChai from 'sinon-chai';
...
+                owner: 'bob',
+                name: 'Im a thing'
+            }
+            let response;
+
+            deviceInstance.ajax.customXHR = this.xhr;
+
+            deviceInstance.updateDevice(patch, true).then((response) => {
+                //console.log(response);
+                expect(patch).to.deep.equal(response);
```
Comment (jontore) [inline test/device.test.js]: What about that it updates the properties on the device?

File: test/user.test.js [hunk 9]
```diff
@@ -5,81 +5,89 @@ import sinon from 'sinon';
...
+            }
 
 
-      userInstance.ajax.customXHR = this.xhr;
-      userInstance.getMyTransmitters().then((result) => {
-        expect(result).to.deep.equal(transmitterStub[0]);
-        done();
-      });
+            userInstance.getUserInfo().then((userInfo) => {
+                expect(userInfo).to.deep.equal(userStub);
```
Comment (jontore) [inline test/user.test.js]: This also updates the properties on the user.

File: tools/connection.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
+export
+default
+
+function connection() {
```
Comment (jontore) [inline tools/connection.js]: Why not use the normal class format here.

File: tools/connection.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    let self = this;
+    this.buffer = [];
+    this.event = function(data) {
```
Comment (jontore) [inline tools/connection.js]: Are we testing this logic anywhere?

File: entities/Device.js (hunk 4 above)
Comment (SmbatYeranyan) [inline entities/Device.js]: Sounds good

File: entities/Device.js (hunk 3 above)
Comment (SmbatYeranyan) [inline entities/Device.js]: Im assuming copy down error :)

File: tools/connection.js (hunk 11 above)
Comment (SmbatYeranyan) [inline tools/connection.js]: added tests for this

**Pull request merged**